### PR TITLE
refactor(plugin-chart-table): migrate to API v1

### PR DESCRIPTION
**Pull request by ktmud**
🏆 Enhancements

Migrate the table viz to use API v1, along with some refactoring related to QueryContext.

1. Add typing for `post_processing`
2. Allow `ChartPlugin` to have custom `ChartProps`
3. Remove `groupby` from `queryFields` as it is [deprecated](https://github.com/apache/superset/pull/9366).
4. Add `GenericDataType`, the chart data type hints for rendering

This PR is a prerequisite of https://github.com/apache/superset/pull/10270

### CODE REVIEW

**Comment by ktmud** [comment]
Blocked by https://github.com/apache-superset/superset-ui/pull/892 because ESLint needs an upgrade in order to fix frivolous lint errors.

File: packages/superset-ui-core/src/query/extractQueryFields.ts
Comment (simchaNielsen) [inline packages/superset-ui-core/src/query/extractQueryFields.ts:66]: can we unit these 2 conditions? like:
```
if (queryMode === QueryMode.aggregate && normalizedKey === 'columns) {
```

File: packages/superset-ui-core/src/query/types/PostProcessing.ts
Comment (simchaNielsen) [inline packages/superset-ui-core/src/query/types/PostProcessing.ts:4]: it should be before all imports

File: packages/superset-ui-core/src/query/types/PostProcessing.ts
Comment (simchaNielsen) [inline packages/superset-ui-core/src/query/types/PostProcessing.ts:62]: should it be some TS type?

File: packages/superset-ui-core/src/utils/removeDuplicates.ts
Comment (simchaNielsen) [inline packages/superset-ui-core/src/utils/removeDuplicates.ts:33]: may be only `Set` enougph? if I got correctly logic do it like (but not sure):
```
items.forEach(x => {
    if (!hash || !hash(x));
      seen.add(x);
    }
  });
return [...seen]
```

File: packages/superset-ui-core/src/query/extractQueryFields.ts
Comment (ktmud) [reply packages/superset-ui-core/src/query/extractQueryFields.ts:66]: Sure. There was more conditions in this branch before but I must have cleaned them up.

File: packages/superset-ui-core/src/query/types/PostProcessing.ts
Comment (ktmud) [reply packages/superset-ui-core/src/query/types/PostProcessing.ts:4]: Right. Thanks for catching this!  The imported were added by VS Code autocomplete and I forgot to check.

File: packages/superset-ui-core/src/query/types/PostProcessing.ts
Comment (ktmud) [reply packages/superset-ui-core/src/query/types/PostProcessing.ts:62]: I think it's find to use string literals here, otherwise we'd need to create a type for each operation, which doesn't seem to add much additional value.

File: packages/superset-ui-core/src/utils/removeDuplicates.ts
Comment (ktmud) [reply packages/superset-ui-core/src/utils/removeDuplicates.ts:33]: Does set spread preserve orders?

Comment (ktmud) [reply packages/superset-ui-core/src/utils/removeDuplicates.ts:33]: [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#iteration_methods) says yes, but this still wouldn't work for my needs because I need to check uniqueness by hash, not the item itself.

Comment (kristw) [reply packages/superset-ui-core/src/utils/removeDuplicates.ts:33]: ideas: can short circuit when `hash` is not used and only check `hash` once.

```ts
function removeDuplicates<T>(items: T[], hash?: (item: T) => unknown): T[] {
  if (hash) {
    const seen = new Set();
    return items.filter(x => {
      const itemHash = hash(x);
      if (seen.has(itemHash)) return false; 
      seen.add(itemHash);
      return true;
    });
  }

  return [...new Set(items)];
}
```

File: plugins/plugin-chart-table/test/tsconfig.json
Comment (kristw) [inline plugins/plugin-chart-table/test/tsconfig.json:6]: Is this intentional?

File: packages/superset-ui-core/src/chart/models/ChartPlugin.ts
Comment (kristw) [inline packages/superset-ui-core/src/chart/models/ChartPlugin.ts:105]: Given we are all in with typescript now. Could also remove the legacy `isRequired` trick that was before the age of typescript. (No need to do in this pr).

**Review by kristw — APPROVED** [review]
Just a few things here and there. Overall lgtm

File: packages/superset-ui-core/src/query/extractQueryFields.ts
Comment (ktmud) [reply packages/superset-ui-core/src/query/extractQueryFields.ts:75]: I try to avoid `switch` nowadays since Python has taught me [`switch` is just a syntactic sugar for if else's](https://stackoverflow.com/questions/46701063/why-doesnt-python-have-switch-case). Plus I hate it when I forgot to call `break`.

File: plugins/plugin-chart-table/test/tsconfig.json
Comment (ktmud) [reply plugins/plugin-chart-table/test/tsconfig.json:6]: Yes as I need to import a contextualized `enzyme` mount from root: https://github.com/ktmud/superset-ui/blob/8e744a77751fc8863522c8938974262e7e3463bd/plugins/plugin-chart-table/test/TableChart.test.tsx#L21

File: packages/superset-ui-core/src/chart/models/ChartPlugin.ts
Comment (ktmud) [reply packages/superset-ui-core/src/chart/models/ChartPlugin.ts:105]: Noted. Let's do this when all major plugins (at their index.js file) are migrated to typescript.

**Review by villebro — APPROVED** [review]
This is amazing stuff, thanks so much for all the improvements here. Let's get this in and deal with any potential issues with higi priority, as I feel the improvements here far outweigh any potential regressions caused by it.

File: packages/superset-ui-core/src/query/types/QueryFormData.ts
Comment (ktmud) [inline packages/superset-ui-core/src/query/types/QueryFormData.ts:157]: Switching order so if something doesn't add up, it will fallback to `SqlaFormData`

File: packages/superset-ui-core/src/utils/removeDuplicates.ts
Comment (ktmud) [reply packages/superset-ui-core/src/utils/removeDuplicates.ts:33]: Good idea! Updated.

File: packages/superset-ui-core/test/utils/removeDuplicates.test.ts
Comment (ktmud) [reply packages/superset-ui-core/test/utils/removeDuplicates.test.ts]: Updated!

File: plugins/plugin-chart-table/src/transformProps.ts
Comment (ktmud) [inline plugins/plugin-chart-table/src/transformProps.ts:77]: I'm just sick and tire of worrying about casing so I choose to use raw form data here. Should probably gradually deprecating this auto-case-transformation for the final charting API as well...

File: plugins/plugin-chart-table/src/types.ts
Comment (ktmud) [inline plugins/plugin-chart-table/src/types.ts:50]: The benefit of not doing case transformation is that we can directly reuse this typing for both `formData` sent to the API and received by the chart renderer (e.g. in both `buildQuery` and `transformProps`.)

**Review by ktmud — COMMENTED** [review]
Adding more comments and replies

**Comment by ktmud** [comment]
> This is amazing stuff, thanks so much for all the improvements here. Let's get this in and deal with any potential issues with higi priority, as I feel the improvements here far outweigh any potential regressions caused by it.

Thank you, @villebro ! I've tested ECharts, NVD3, Box plot, Word Cloud and tried to be as comprehensive as I can. Also wanted to add more Cypress tests for the missing chart types (boxplot, time series table, for example) before merging, but didn't have enough time since a lot of my time has been spent on [updating Cypress itself](https://github.com/apache/superset/pull/12605)..

It would be super nice if someone from Preset could prioritize adding E2E tests for viz types or completing existing ones with more use cases so we can have more confidence in doing bigger viz plugin improvements like this. 

cc @junlincc

**Comment by ktmud** [comment]
I'm going to a new version and update https://github.com/apache/superset/pull/10270

**Pull request merged**
